### PR TITLE
MH-12709: Replacing broken conflict detection logic

### PR DESCRIPTION
**Pull request by gregorydlogan**
The previous check searched for events which started or ended in a pair of windows.  This missed several edge cases, and was thus replaced.

*Work sponsored by UCT*

### CODE REVIEW

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java [hunk 1]
```diff
@@ -1207,6 +1207,12 @@ private boolean isBlacklisted(String mediaPackageId, Date start, Date end, Strin
     return searchInternal(startsFrom, startsTo, endFrom, endTo, alreadyScheduledEvents);
   }
 
+  /*
+   * Returns a list of events which start and/or end within a given date range.  So you could search for things which start
+   * on Tuesday between 0900 and 1000, and end between 1500 and 1600, and you could get an event which started at Epoch and
+   * ended at 1559 on Tuesday.  This is *NOT* appropriate for conflict checking, and does not check for any edge cases.
+   * sUse checkScheduleConflicts instead.
```
Comment (smarquard) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: sUse > Use

Comment (gregorydlogan) [reply modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: That's there that s went.  I was wondering!

**Comment by smarquard** [comment]
Can you give some examples of the edge cases that were missed by the old code and are caught by the new code? For testing purposes, as there aren't unit tests that check this.

**Comment by gregorydlogan** [comment]
Why don't I write up some unit tests too while I'm at it then.  I checked the coverage which seemed good, but that can be misleading I suppose.

Edge cases, some of which are likely tested for:
- Event A contains event B entirely
- Event A is contained by event B entirely
- Event A begins before event B, but ends during event B -> This was not caught originally
- Event A begins during event B, but ends after event B ends -> This was not caught originally
- Event A begins at the same time as event B (to the second)
- Event A ends at the same time as event B (to the second)

Unit tests likely won't come today, but hopefully soon(tm).

**Comment by gregorydlogan** [comment]
Ok, I lied, quick tests built today.

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java [hunk 2]
```diff
@@ -1232,6 +1238,32 @@ public int compare(ARecord o1, ARecord o2) {
...
 
+  /*
+   * Returns a list of conflicting mediapackages.  This method checks for containment, starts-during, and ends-during.
+   */
+  private List<MediaPackage> checkScheduleConflicts(Date checkStart, Date checkEnd, ARecord[] records) {
+    final List<ARecord> result = new ArrayList<>();
+    for (final ARecord r : records) {
+      final Date start = r.getProperties().apply(Properties.getDate(START_DATE_CONFIG));
+      final Date end = r.getProperties().apply(Properties.getDate(END_DATE_CONFIG));
+      if (checkStart.equals(start) || checkEnd.equals(end) ||      //If the start or end times are identical (.after and .before are strict comparisons)
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: I think you have missed checkEnd.equals(start) (the potential event ends at the same time when a new event starts) and checkStart.equals(end) (the potential event starts at the same time when a new event finished). Or is this legal?

**Review by staubesv — CHANGES_REQUESTED** [review]
I've done with the first round of review. Depending on the definition of scheduling conflict, the code seems to miss two corner cases.

I think those are conflicts, aren't they?

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java [hunk 3]
```diff
@@ -1232,6 +1238,32 @@ public int compare(ARecord o1, ARecord o2) {
...
+          checkStart.after(start) && checkStart.before(end) ||     //If the potential event starts during event r
+          checkEnd.before(end) && checkEnd.after(start) ||         //If the potential event ends during event r
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: checkend.after(start) && checkEnd.before(end) might be easier to read (as more "left to right")

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java [hunk 4]
```diff
@@ -1207,6 +1207,12 @@ private boolean isBlacklisted(String mediaPackageId, Date start, Date end, Strin
     return searchInternal(startsFrom, startsTo, endFrom, endTo, alreadyScheduledEvents);
   }
 
+  /*
+   * Returns a list of events which start and/or end within a given date range.  So you could search for things which start
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: More than 120 characters per line

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java (hunk 2 above)
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: More than 120 characters per line

Comment (smarquard) [reply modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: I think there should be a minimum separation enforced between end of one event and start of the next, probably a minute. So ends 9:59am starts 10am is legal, but ends 10am starts 10am is not.

Comment (staubesv) [reply modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: @gregorydlogan Could you please add the two corner cases mentioned above and the unit tests for those two cases?

Comment (gregorydlogan) [reply modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: @smarquard @staubesv This isn't actually part of the CA API, but I'm not sure that it shouldn't be.  I'll add it to the notes, and to the unit tests.

Comment (staubesv) [reply modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: @gregorydlogan Let's assume that our CA API is underspecified: It doesn't make sense to schedule events where Event A starts just 2 seconds after Event B on the same capture agent if that capture agent cannot do this. I think the future CA API should say some words about how fast a capture agent is supposed to be able to stop an recording and start the next one. Probably, it will require at least "some" time to do so... but we would need to ask the capture agent vendor for what is a realistic value here (that can be guaranteed by the CA with reasonable effort).

Update: Probably, it would make sense that the future CA would support retrieving such information from the CA itself, as the "quickness" might be specific to CAs. If the CA reports the "minimum pause required", Opencast could consider that value per CA when doing scheduling... and vendors could be motivated to be "quicker" than others ;-)

**Comment by gregorydlogan** [comment]
Changes made.

File: modules/matterhorn-scheduler-impl/src/test/java/org/opencastproject/scheduler/impl/SchedulerServiceImplTest.java [hunk 5]
```diff
@@ -1895,6 +1895,12 @@ public void testFindConflictingEvents() throws Exception {
       //Event A contains event B entirely
       conflicts = schedSvc.findConflictingEvents("Device A", new Date(currentTime + hours(23)), new Date(currentTime + hours(26)));
       assertEquals(1, conflicts.size());
+
+      //Event A ends with less than one minute before event B starts
+      conflicts = schedSvc.findConflictingEvents("Device A", new Date(currentTime + hours(23)), new Date(currentTime + hours(24) - seconds(1)));
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/test/java/org/opencastproject/scheduler/impl/SchedulerServiceImplTest.java]: The assertion is missing

File: modules/matterhorn-scheduler-impl/src/test/java/org/opencastproject/scheduler/impl/SchedulerServiceImplTest.java [hunk 6]
```diff
@@ -1895,6 +1895,12 @@ public void testFindConflictingEvents() throws Exception {
...
+
+      //Event A begins than one minute after event B ends
+      conflicts = schedSvc.findConflictingEvents("Device A", new Date(currentTime + hours(25) + seconds(1)), new Date(currentTime + hours(27)));
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/test/java/org/opencastproject/scheduler/impl/SchedulerServiceImplTest.java]: The assertion is missing

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java [hunk 7]
```diff
@@ -1246,10 +1249,17 @@ public int compare(ARecord o1, ARecord o2) {
...
-          checkStart.after(start) && checkStart.before(end) ||     //If the potential event starts during event r
-          checkEnd.before(end) && checkEnd.after(start) ||         //If the potential event ends during event r
-          checkStart.before(start) && checkEnd.after(end)) {       //If the potential event begins before, and ends after event r (ie, containment)
+      /*
+      If the start or end times are identical (.after and .before are strict comparisons) OR
+      If the potential event starts during event r OR
+      If the potential event ends during event r OR
+      If the potential event begins before, and ends after event r (ie, containment)
+      */
+      if (checkStart.equals(start) || checkEnd.equals(end)
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: This test now seems redundant: The test eventWithinMinimumSeparation(...) already does test this, doesn't it?

File: modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java [hunk 8]
```diff
@@ -1232,6 +1241,52 @@ public int compare(ARecord o1, ARecord o2) {
...
+    });
+    return Stream.mk(result).bind(recordToMp).toList();
+  }
+
+  /**
+   * Returns true of checkStart is within EVENT_MINIMUM_SEPARATION_SECONDS of either the start or end dates, or checkEnd
+   * is within EVENT_MINIMUM_SEPARATION_SECONDS of either the start or end dates.  False otherwise
+   */
+  private boolean eventWithinMinimumSeparation(Date checkStart, Date checkEnd, Date start, Date end) {
+    if (Math.abs(checkStart.getTime() - start.getTime()) <= EVENT_MINIMUM_SEPARATION_MILLISECONDS
```
Comment (staubesv) [inline modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java]: @gregorydlogan One last thing: Please change the '<=' in this tests to '<' for the following reason:

The Admin UI allows scheduling start, end and duration by minutes. If a user now wants Recording 1 (3:00PM-3:29PM, duration = 29 minutes) and Recording 2 (3:30PM-4:00PM), this will be reported as a conflict which is unclear to the user as from the data he sees and knows, there is no conflict here

If you change the "<=" to "<", that won't happen which improves usability quite a lot. Probably requires to adjust some comments.

Also, 3:29 to 3:30 has 60 seconds difference with means EVENT_MINIMUM_SEPARATION_SECONDS should be ok.

**Review by staubesv — APPROVED** [review]
